src/main/scala/tsiHost/TsiHostWidget.scala

```diff
@@ -0,0 +1,335 @@
+ *
+ * @param tsiHostParams the base params for the widget
+ * @param controlBus the bus that will be the client for the mmio
+ */
+case class TSIHostWidgetAttachParams(
+  tsiHostParams: TSIHostParams,
+  controlBus: TLBusWrapper)(implicit val p: Parameters)
+
+/**
+ * Factory object to create TSI Host Widgets
+ */
+object TLTSIHostWidget {
+  /**
+   * Just create a TSI widget and connect it to the specified bus
+   *
+   * @param attachParams params to connect the widget to the mmio bus and instantiate it
+   * @return a TSI Host Widget and a TL node to connect to backing memory
+   */
+  def attach(attachParams: TSIHostWidgetAttachParams): (TLTSIHostWidget, TLIdentityNode) = {
+    implicit val p = attachParams.p
+
+    val name = "tsi_widget"
+    val cbus = attachParams.controlBus
+
+    // create a TL TSI Host Widget
+    val tsiHostWidget = LazyModule(new TLTSIHostWidget(cbus.beatBytes, attachParams.tsiHostParams))
+    tsiHostWidget.suggestName(name)
+
+    // connect the manager (mmioNode) to the client (the bus wanted)
+    cbus.coupleTo(s"slave_named_$name") {
+      tsiHostWidget.mmioNode := TLFragmenter(cbus.beatBytes, cbus.blockBytes) := _
+    }
+
+    // connect the clock and reset
+    InModuleBody { tsiHostWidget.module.clock := cbus.module.clock }
+    InModuleBody { tsiHostWidget.module.reset := cbus.module.reset }
+
+    // expose the widget and the client node
+    (tsiHostWidget, tsiHostWidget.externalClientNode)
+  }
+
+  /**
+   * Create a TSI Host Widget, attach it to the specified bus, then create an i/o to connect to
+   *
+   * @param attachParams params to connect with the widget to the bus
+   * @param memoryBus the memory bus to connect the widget to
+   * @return i/o to connect to the tsi widget
+   */
+  def attachAndMakePort(attachParams: TSIHostWidgetAttachParams, memoryBus: TLBusWrapper): ModuleValue[TSIHostWidgetIO] = {
+    implicit val p = attachParams.p
+
+    val (tsiHost, tsiHostMemNode) = attach(attachParams)
+
+    // connect the widget mem node to the input membus
+    memoryBus.coupleFrom(s"master_named_${tsiHost.name}") {
+      _ := tsiHostMemNode
+    }
+
+    // create an io node that automatically makes an IO based on a bundle (can either make it a source or a sink)
+    val tsiHostNode = BundleBridgeSource(() => new TSIHostWidgetIO(attachParams.tsiHostParams.serialIfWidth))
+    InModuleBody { tsiHostNode.makeIO()(ValName(tsiHost.name)) }
+  }
+
+  /**
+   * Tieoff all inputs to the widget
+   *
+   * @param port the i/o bundle interfacing with the widget
+   * @param clock the clock to run the widget on
+   * @return none
+   */
+  def tieoff(port: TSIHostWidgetIO, clock: Clock) {
+    port.serial.in.bits := 0.U
+    port.serial.in.valid := 0.U
+    port.serial.out.ready := 0.U
+    port.serial_clock := clock
+  }
+
+  /**
+   * Connect the output to the input
+   *
+   * @param port the i/o bundle to connect
+   * @return none
+   */
+  def loopback(port: TSIHostWidgetIO) {
```

Comment:
👍 cool. Can this create live lock if you write to this widget with its self?

Reply:
Maybe. I haven't tried it. Frankly, I don't use this in the current unit tests.